### PR TITLE
add fine grained cache control header support

### PR DESCRIPTION
**Pull request by danielpoe**
Currently there is only a SetNoCache Helper for responses

Issues:
- No default cache control strategy can be set
- The current "SetNoCache" - is not 100% correct 
- There is always deeper caching know how involved for setting correct cache related http headers - we should provide a consitent and easy way to set and only "build" reasonable header combinations

### CODE REVIEW

File: framework/web/result.go [hunk 1]
```diff
@@ -75,16 +78,44 @@ type (
 		RenderResponse
 		Error error
 	}
+
+	//CacheControlHeader - holds the possible directives for Cache Control Headers
```
[the remark this answers is not in the document]
Comment (danielpoe) [reply framework/web/result.go]: In the method SetHeaders

no-cache and ETag is not mutually exclusive by the way

no-store and ETag is

File: framework/web/result.go [hunk 2]
```diff
@@ -384,3 +429,114 @@ func (r *Responder) getLogger() flamingo.Logger {
...
+	if c.Visibility == "private" {
+		cacheControlValues = append(cacheControlValues, "private")
+	}
+
+	if len(cacheControlValues) > 0 {
+		header.Set("Cache-Control", strings.Join(cacheControlValues, ", "))
+	}
+}
+
+//NewCacheStrategy returns new cache Strategy
```
Comment (thebod) [inline framework/web/result.go]: This factory is not necessary, and actually not always used, e.g. by the Filter.

File: framework/web/filter/cachestrategy.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+import (
+	"context"
+	"flamingo.me/flamingo/v3/framework/web"
+	"net/http"
+)
+
+type (
+	//CacheStrategy - a filter that sets CacheDirective if not present
+	CacheStrategy struct {
```
Comment (thebod) [inline framework/web/filter/cachestrategy.go]: Make this type private

File: framework/web/filter/cachestrategy.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	response := chain.Next(ctx, r, w)
+	if defaultResponse, ok := response.(*web.Response); ok {
+		if defaultResponse.CacheDirectives == nil && r.Request().Method == http.MethodGet {
+			cacheStrategy := web.CacheStrategy{}
+			cacheStrategy.SetIsReusable(f.DefaultIsReuseable)
+			cacheStrategy.SetRevalidateEachTime(f.DefaultRevalidateEachTime)
+			cacheStrategy.SetAllowIntermediateCaches(f.DefaultAllowIntermediateCaches)
+			cacheStrategy.SetMaxCacheLifetime(int(f.DefaultMaxCacheLifetime))
+			defaultResponse.CacheDirectives = cacheStrategy.Build()
+			return defaultResponse
```
Comment (thebod) [inline framework/web/filter/cachestrategy.go]: Return the `response` variable. The `defaultResponse` does not contain everything, you basically unpack the `web.Response` from the actual `web.Result`.

File: framework/web/result.go [hunk 5]
```diff
@@ -75,16 +77,58 @@ type (
...
+	CacheStrategy struct {
+		isReusable bool
+		revalidateEachTime bool
+		allowIntermediateCaches bool
+		maxCacheLifetime int
+		etag string
+	}
+
+	//CacheDirectives - holds the possible directives for Cache Control Headers and other Http Caching
+	CacheDirectives struct {
```
Comment (thebod) [inline framework/web/result.go]: `CacheDirective` singular

File: framework/web/filter/cachestrategy.go [hunk 6]
```diff
@@ -17,18 +18,26 @@ type (
...
-			cacheStrategy.SetAllowIntermediateCaches(f.DefaultAllowIntermediateCaches)
-			cacheStrategy.SetMaxCacheLifetime(int(f.DefaultMaxCacheLifetime))
-			defaultResponse.CacheDirectives = cacheStrategy.Build()
-			return defaultResponse
-		}
+	if r.Request().Method != http.MethodGet {
+		return response
+	}
+	switch typedResponse := response.(type) {
+	case *web.RenderResponse:
```
Comment (thebod) [inline framework/web/filter/cachestrategy.go]: Is this supposed to be `*web.RenderResponse`? You might want to use `*web.Response`?

**Pull request merged**